src/main/java/seedu/address/logic/parser/HintParser.java

```diff
@@ -0,0 +1,318 @@
+        case ReviewsCommand.COMMAND_WORD:
+            return generateHintForIndexedCommand(" view book review");
+        case SearchCommand.COMMAND_WORD:
+            return generateSearchHint();
+        case SelectCommand.COMMAND_WORD:
+            return generateHintForIndexedCommand(" select a book");
+        case ThemeCommand.COMMAND_WORD:
+            return generateThemeHint();
+        case UndoCommand.COMMAND_WORD:
+            return " undo last modification";
+        default:
+            return "";
+        }
+    }
+
+    //@@author
+    /**
+     * Parses the end of arguments to check if user is currently typing a prefix that is in prefixes.
+     * Returns hint if user is typing a prefix.
+     * Returns empty Optional if user is not typing a prefix.
+     */
+    private Optional<String> generatePrefixHintBasedOnEndArgs(Prefix... prefixes) {
+        for (Prefix p : prefixes) {
+            String prefixString = p.getPrefix();
+            String parameter = prefixIntoParameter(p);
+            for (int i = 1; i <= prefixString.length(); ++i) {
+                if (arguments.endsWith(" " + prefixString.substring(0, i))) {
+                    return Optional.of(prefixString.substring(i, prefixString.length()) + parameter);
+                }
+            }
+        }
+        return Optional.empty();
+    }
+
+    /**
+     * Parses arguments to check for parameters that have not been used.
+     * Only the specified {@code prefixes} will be checked.
+     * Returns hint for all parameter that are not present.
+     * Returns {@code defaultHint} if all parameters are present.
+     */
+    private String showUnusedParameters(String defaultHint, String preamble, Prefix... prefixes) {
+        String whitespace = userInput.endsWith(" ") ? "" : " ";
+        ArgumentMultimap argumentMultimap = ArgumentTokenizer.tokenize(arguments, prefixes);
+
+        StringBuilder sb = new StringBuilder();
+        if (preamble != null && argumentMultimap.getPreamble().trim().isEmpty()) {
+            sb.append("[").append(preamble).append("] ");
+        }
+
+        for (Prefix p : prefixes) {
+            Optional<String> parameterOptional = argumentMultimap.getValue(p);
+            if (!parameterOptional.isPresent()) {
+                sb.append("[").append(p.getPrefix()).append(prefixIntoParameter(p)).append("] ");
+            }
+        }
+        if (sb.length() == 0) {
+            return whitespace + defaultHint;
+        }
+        return whitespace + sb.toString();
+    }
+
+    //@@author fishTT
+    /**
+     * Currently this method is always called after generatePrefixHintBasedOnEndArgs.
+     * It parses arguments to check for parameters that have not been filled up.
+     * Only the specified {@code prefixes} will be checked.
+     * Returns hint for parameter that is not present.
+     * Returns {@code defaultHint} if all parameters are present.
+     */
+    private String offerHint(String defaultHint, Prefix... prefixes) {
```

Comment:
can make it clearer how this is different from previous method?

Reply:
the second is to check if all parameters are present